### ISSUE TITLE
[Inductor] `F.hardtanh` compiled backward silently drops gradient at bf16 boundary values

### ISSUE THREAD

**Issue by wuyii8941** [pytorch/pytorch#185472]
## 🐛 Describe the bug

`torch.compile(F.hardtanh)` backward pass produces incorrect gradients for bfloat16 tensors when the input value equals the bf16 representation of the bound. The compiled backward zeros the gradient where eager correctly passes it through.

This can silently corrupt training when using `hardtanh` activation with bf16 (e.g., AMP or bf16 training). Roughly half of all non-exact bound values are affected.

## Minimal reproducer

```python
import torch
import torch.nn.functional as F

# bf16(0.7) = 0.69921875 (bf16 rounds 0.7 down)
x = torch.tensor([0.69921875], dtype=torch.bfloat16, device='cuda', requires_grad=True)

# Eager: gradient passes through (value is within bounds)
F.hardtanh(x, min_val=-0.7, max_val=0.7).backward()
print(f"eager grad:    {x.grad}")  # tensor([1.], ...) ← correct

# Compiled: gradient is zeroed
x2 = torch.tensor([0.69921875], dtype=torch.bfloat16, device='cuda', requires_grad=True)
torch._dynamo.reset()
torch.compile(lambda y: F.hardtanh(y, min_val=-0.7, max_val=0.7))(x2).backward()
print(f"compiled grad: {x2.grad}")  # tensor([0.], ...) ← WRONG
```

**Expected**: `x2.grad = tensor([1.])` (same as eager — value is within the hardtanh bounds).  
**Actual**: `x2.grad = tensor([0.])` (gradient incorrectly dropped).

fp32 is NOT affected — both eager and compiled give `grad=1.0` for the same input value.

## Scope

This is not a single-value edge case. Testing all bound values in `[0.01, 9.99]` with step 0.01:
- **477 out of ~930 non-exact bound values trigger the bug** (any bound whose bf16 representation rounds DOWN)
- Both upper and lower boundaries are affected
- `torch.clamp` backward is NOT affected (different codepath)
- `relu6` is NOT affected (bounds 0 and 6 are exactly representable)

## Root cause

The `hardtanh_backward` decomposition in `torch/_decomp/decompositions.py` (line 212):

```python
@register_decomposition(aten.hardtanh_backward)
@out_wrapper("grad_input")
def hardtanh_backward(grad_output, self, min_val, max_val):
    return torch.where((self <= min_val) | (self >= max_val), 0.0, grad_output)
```

The issue is a semantic mismatch between the decomposition and the eager CUDA kernel at the dtype boundary:

For `x = bf16(0.7) = 0.69921875`, `max_val = 0.7`:

| | Comparison | Result | Gradient |
|---|---|---|---|
| **Eager** (CUDA kernel) | `bf16(x) < float(0.7)` → `0.6992 < 0.7` | within bounds | **1.0** ✓ |
| **Compiled** (decomposition) | `bf16(x) >= bf16(0.7)` → `0.6992 >= 0.6992` | at boundary | **0.0** ✗ |

The eager CUDA kernel compares the bf16 input against the original float scalar, while the compiled decomposition comparison effectively uses bf16 precision for the bound, causing boundary values to be misclassified as "clamped".

## Versions

- PyTorch: 2.13.0.dev20260521+cu126
- GPU: NVIDIA RTX A6000 (sm_86)
- CUDA: 12.6 (bundled)
- Python: 3.11
- OS: Linux

cc @chauhang @penguinwu

**Comment by Lawson-Darrow** [comment]
Took a look and confirmed it reproduces, and not only on CUDA: on CPU with the inductor backend (torch 2.12), compiled `hardtanh` backward returns grad 0.0 for `x = bf16(0.7) = 0.69921875` with bounds ±0.7, while eager returns 1.0.

Root cause is what you describe. The `hardtanh_backward` decomp does `self >= max_val` with `self` in bf16, so the float bound effectively rounds to bf16 and the boundary value is misclassified as clamped. Eager keeps the comparison in higher precision.

The fix is to run the boundary comparison in higher precision (upcast bf16/fp16 `self` to float32 before the `where`), which is the opposite direction from the cast-scalar-to-dtype fix used for `threshold`/`hardshrink`. I patched the decomp that way locally and the compiled grad then matches eager (1.0). Happy to send a PR if you can mark this actionable.
